### PR TITLE
Added SHA256 in real time

### PR DESCRIPTION
**Pull request by duyvan1995**
The SHA256 firmware is in CAN_Logger_with_Autobaid_and_Requests_no_CAN2_AES_CBC

### CODE REVIEW

**Comment by jeremy-daily** [comment]
Here are some test vectors:
```
HashSHA256Test sha256_test[] =
{
  {(unsigned char*)"",{0xe3b0c442,0x98fc1c14,0x9afbf4c8,0x996fb924,0x27ae41e4,0x649b934c,0xa495991b,0x7852b855}},
  {(unsigned char*)"abc",{0xba7816bf,0x8f01cfea,0x414140de,0x5dae2223,0xb00361a3,0x96177a9c,0xb410ff61,0xf20015ad}},
  {(unsigned char*)"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq",{0x248d6a61,0xd20638b8,0xe5c02693,0x0c3e6039,0xa33ce459,0x64ff2167,0xf6ecedd4,0x19db06c1}},
  {(unsigned char*)"abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu",{0xcf5b16a7,0x78af8380,0x036ce59e,0x7b049237,0x0b249b11,0xe8f07a51,0xafac4503,0x7afee9d1}},
  {(unsigned char*)"This is exactly 64 bytes long, not counting the terminating byte",{0xab64eff7,0xe88e2e46,0x165e29f2,0xbce41826,0xbd4c7b35,0x52f6b382,0xa9e7d3af,0x47c245f8}},
  {(unsigned char*)"For this sample, this 63-byte string will be used as input data",{0xf08a78cb,0xbaee082b,0x052ae070,0x8f32fa1e,0x50c5c421,0xaa772ba5,0xdbb406a2,0xea6be342}},
  {(unsigned char*)"And this textual data, astonishing as it may appear, is exactly 128 bytes in length, as are both SHA-384 and SHA-512 block sizes",{0x0ab80334,0x4830f920,0x89494fb6,0x35ad00d7,0x6164ad6e,0x57012b23,0x7722df0d,0x7ad26896}},
  {(unsigned char*)"By hashing data that is one byte less than a multiple of a hash block length (like this 127-byte string), bugs may be revealed.",{0xe4326d04,0x59653d7d,0x3514674d,0x713e74dc,0x3df11ed4,0xd30b4013,0xfd327fdb,0x9e394c26}},
  {(unsigned char*)"The quick brown fox jumps over the lazy dog",{0xd7a8fbb3,0x07d78094,0x69ca9abc,0xb0082e4f,0x8d5651e4,0x6d3cdb76,0x2d02d0bf,0x37c9e592}},
  {(unsigned char*)"This test tries to use the n-block utility from the hash library and as a matter of fact we're trying to get only 128 characters",{0x2ce675bd,0x3b70e104,0xd696d1b2,0x5bf3d42b,0x2b45cd77,0x6d4f590f,0x210f12c4,0x4bf473d5}},
};
```

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 1]
```diff
@@ -262,8 +260,78 @@ bool stream = false;
...
+    while((RNG_SR & RNG_SR_OREG_LVL(0xF)) == 0); // wait
+    return RNG_OR;
+}
+/* K66F  CAU tests   SHA256  AES-CBC
+crypto assist co-processor
+*/
+#include "CryptoAccel.h" //Makes it arduino compatible
+#include "cau_api.h"
+    
+unsigned char cipher_text[512];
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: here (see below)

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 2]
```diff
@@ -262,8 +260,78 @@ bool stream = false;
...
+    iv_and_key[4*m+2] = (random_number & 0x0000FF00) >> 8;
+    iv_and_key[4*m+3] = (random_number & 0x000000FF);
+    }
+}
+
+//AES CBC funtion
+void aes_cbc_encrypt(const unsigned char *data, unsigned char *cipher_text){
+  //Data length should be a multiple of 16 bytes
+  // Need to initialize out with the initialization vector
+  for (uint32_t j=0; j < 512; j+=16){
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: this `512` is a magic number for the maximum size that an plaintext can be in your logging scheme. You should at least refactor this as a well-named constant. Ideally derive the value 512 (i.e. it is the buffer size, which is itself chosen b/c ...) and also comment the rationale for it.

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 3]
```diff
@@ -262,8 +260,78 @@ bool stream = false;
...
+unsigned int hash_counter;
+#define SHA256_BLOCK_SIZE 32            // SHA256 outputs a 32 byte digest
+BYTE hash_text[64];
+BYTE hash[SHA256_BLOCK_SIZE];
+BYTE before_closing_hash[512];
+boolean first_buffer_sent;
+Sha256* sha256Instance;
+
+// random RNG
+#define REPS 50
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:272]: Some comment rationale is needed here. This isn't just "random RNG". Consider answering the questions: why `REPS` of `50`? why not any other value? has this been tested? Is this the same approach used somewhere else in a trusted source (e.g. boringSSL, bearSSL, NaCL, some paper...)

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 4]
```diff
@@ -262,8 +260,78 @@ bool stream = false;
...
+
+#define RNG_CR_GO_MASK                           0x1u
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: This series of defines comes from somewhere else (the same strings show up in Teensy projects on the internet). You should mention where they come from I think.

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 5]
```diff
@@ -262,8 +260,78 @@ bool stream = false;
...
+unsigned char cipher_text[512];
+unsigned char aeskey[16], keysched[4 * 44], in[16], out[16], init_vector[16];
+
+//Generate ranndom iv and key function, 32 bytes number
+void iv_key_RNG(){
+    
+    for (int m =0;m<8;m++){
+    SIM_SCGC6 |= SIM_SCGC6_RNGA; // enable RNG
+    RNG_CR &= ~RNG_CR_SLP_MASK;
+    RNG_CR |= RNG_CR_HA_MASK;  // high assurance, not needed
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:303]: why is high assurance not needed?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 6]
```diff
@@ -338,41 +424,35 @@ void check_buffer(){
     
     data_buffer[491] = Can0.readREC();
     data_buffer[492] = Can1.readREC();
-    data_buffer[493] = uint8_t(Can2.errorCountRX());
+    //data_buffer[493] = uint8_t(Can2.errorCountRX());
+    data_buffer[493] = 0;
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: looks like you are committing test/debug changes to zero the rx error count here

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 7]
```diff
@@ -338,41 +424,35 @@ void check_buffer(){
...
     
     data_buffer[494] = Can0.readTEC();
     data_buffer[495] = Can1.readTEC();
-    data_buffer[496] = uint8_t(Can2.errorCountTX());
+    //data_buffer[496] = uint8_t(Can2.errorCountTX());
+    data_buffer[496] = 0;
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: looks like you are committing test/debug changes to zero the tx error count here

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 8]
```diff
@@ -784,8 +883,8 @@ void setup(void) {
   digitalWrite(SILENT_2,LOW);
 
   // Setup chip select pin for the MCP2515
-  pinMode(CS_CAN, OUTPUT);
-  digitalWrite(CS_CAN,HIGH);
+ // pinMode(CS_CAN, OUTPUT);
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 9]
```diff
@@ -812,9 +911,15 @@ void setup(void) {
   button.setClickTicks(500);
   
   // Select the CAN on J1939 Pins F and G
-  pinMode(CAN_SWITCH,OUTPUT);
-  digitalWrite(CAN_SWITCH,HIGH);
-
+  //pinMode(CAN_SWITCH,OUTPUT);
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:906]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 10]
```diff
@@ -835,9 +940,9 @@ void setup(void) {
   txmsg.len = 8;
 
   // Setup MCP CAN
-  if(Can2.begin(MCP_ANY, CAN_250KBPS, MCP_16MHZ) == CAN_OK) Serial.println("MCP2515 Initialized Successfully!");
-  else Serial.println("Error Initializing MCP2515...");
-  Can2.setMode(MCP_NORMAL);
+  //if(Can2.begin(MCP_ANY, CAN_250KBPS, MCP_16MHZ) == CAN_OK) Serial.println("MCP2515 Initialized Successfully!");
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 11]
```diff
@@ -835,9 +940,9 @@ void setup(void) {
...
+  //else Serial.println("Error Initializing MCP2515...");
+  //Can2.setMode(MCP_NORMAL);
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 12]
```diff
@@ -916,7 +1043,7 @@ void rx_message_routine(uint32_t RXCount){
   if ((rxmsg.id & CAN_ERR_FLAG) == CAN_ERR_FLAG){
     if (current_channel == 0) ErrorCount0++;
     else if (current_channel == 1) ErrorCount1++;
-    else if (current_channel == 2) ErrorCount2++;
+    //else if (current_channel == 2) ErrorCount2++;
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 13]
```diff
@@ -930,7 +1057,7 @@ void rx_message_routine(uint32_t RXCount){
       txmsg.buf[2] = 0x00; // Separation time in milliseconds
       if (current_channel == 0) send_Can0_message(txmsg);
       else if (current_channel == 1) send_Can1_message(txmsg);
-      else if (current_channel == 2) send_Can2_message(txmsg);
+      //else if (current_channel == 2) send_Can2_message(txmsg);
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 14]
```diff
@@ -947,14 +1074,14 @@ void loop(void) {
     current_channel = 1;
     rx_message_routine(RXCount1);
   }
-  if (Can2.readMsgBuf(&rxId, &ext_flag, &len, rxBuf) == CAN_OK){
+  /*if (Can2.readMsgBuf(&rxId, &ext_flag, &len, rxBuf) == CAN_OK){
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino [hunk 15]
```diff
@@ -976,7 +1103,7 @@ void loop(void) {
         txmsg.buf[2] = (request_pgn[request_index] & 0xFF0000) >> 16; //These are in reverse byte order.
         send_Can0_message(txmsg);
         if (RXCount1 > 0) send_Can1_message(txmsg);
-        if (RXCount2 > 0) send_Can2_message(txmsg);
+        //if (RXCount2 > 0) send_Can2_message(txmsg);
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: more test/debug code commits? (many more instances follow that I won't continue to highlight)

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CryptoAccel.h [hunk 16]
```diff
@@ -0,0 +1,31 @@
+#ifndef CryptoAccel_h_
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CryptoAccel.h]: where does this header come from and what is its license?

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/cau_api.h [hunk 17]
```diff
@@ -0,0 +1,390 @@
+/*
```
Comment (BenGardiner) [inline CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/cau_api.h]: where does this header come from and what is its license?

**Review by BenGardiner — CHANGES_REQUESTED** [review]
Also:
* don't commit generates/compiled files such as `AN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/lib_mmcau.a` and `CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/mmcau_aes_functions.S` . I suggest adding `*.a` and `*.S` to the `.gitignore` file.

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 13 above)
Comment (duyvan1995) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: We replaced the CAN controller for the third CAN channel CAN2 so a new library is needed to be able to use CAN2. I commented them out for now.

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 4 above)
Comment (duyvan1995) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: This RNG function will be replaced by the RNG from the hardware security module ATECC608A in the future

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 6 above)
Comment (duyvan1995) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: There is no CAN2 so I set this information to 0 for now

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 7 above)
Comment (duyvan1995) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: There is no CAN2 so I set this information to 0 for now

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 10 above)
Comment (duyvan1995) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: We replaced the CAN controller for the third CAN channel CAN2 so a new library is needed to be able to use CAN2. I commented them out for now.

**Comment by jeremy-daily** [comment]
> 
> 
> Also:
> 
>     * don't commit generates/compiled files such as `AN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/lib_mmcau.a` and `CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/mmcau_aes_functions.S` . I suggest adding `*.a` and `*.S` to the `.gitignore` file.

@BenGardiner The assembly files are the only available source for the working with the crypto. I'll look in the license for these source files.

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CryptoAccel.h (hunk 16 above)
Comment (jeremy-daily) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CryptoAccel.h]: https://github.com/PaulStoffregen/CryptoAccel/blob/master/Freescale_License.txt

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 2 above)
Comment (jeremy-daily) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino]: I changed these 512 values to a define for the BUFFER_SIZE

File: CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino (hunk 3 above)
Comment (jeremy-daily) [reply CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:272]: I'm not sure on this. @duyvan1995 where did this RNG code come from?

**Comment by jeremy-daily** [comment]
I have been working on a different fork for this code: 
https://github.com/SystemsCyber/CAN-Logger-3
I've addressed some of the issues @BenGardiner brought up in that fork. I suggest we finish this pull request and move over to the new fork and keep working. There are a handful of features we need to get worked out: 
1. Provisioning,
2. J1708
3. Using the ATECC for RNGs
4. Digitally signing the encrypted binary files.
5. Digtally signing the meta_data file
6. Separating the meta_data for each binary file into their own signed files.

**Review by jeremy-daily — APPROVED** [review]

**Comment by BenGardiner** [comment]
ok. let's move over there

**Pull request merged**